### PR TITLE
scene_get_hierarchy: paginate plugin-side instead of shipping the whole tree

### PR DESCRIPTION
**Pull request by eth0net**
## Motivation

`get_scene_tree` ignored `offset`/`limit`: the plugin always walked the entire tree, built a node dict + clean scene path for **every** node, and shipped the whole thing over the WebSocket — then `scene_get_hierarchy` sliced it in Python. So `limit=1` on a large scene still paid the full walk + serialize + transport cost.

## Change

Thread `offset`/`limit` into the plugin. The walk still visits every node for an accurate `total_count`, but only nodes inside the `[offset, offset+limit)` window are materialized (dict + `McpScenePath.from_node`) and shipped. `limit <= 0` means unlimited (the hierarchy resource reads the whole tree).

Python side passes `offset`/`limit`/`depth` through and returns the plugin's paginated result when it carries `has_more`; otherwise it falls back to slicing the full list. That keeps a version-skewed **new-server / old-plugin** pair correct and leaves existing behavior intact. Also drops the dead `root` field — the real plugin never sent it, so it was always `""` in production.

## Tests

- GDScript: pagination window, offset, `total_count`, offset-past-end, and unlimited (`test_scene.gd`).
- Python: the fallback-slice path (old plugin shape) **and** the plugin-paginated passthrough path; updated the workflow readback to verify the loaded scene via the node list rather than the removed `root`.
- Rebased onto v3.0.2; parse-clean; Python suite green locally (the two `test_cli_reload` failures are a local port-9500-in-use artifact, unrelated). Full Godot `test_run` runs in CI.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added pagination to scene hierarchy results via `offset` and `limit`.
  * Pagination now includes `total_count`, `offset`, `limit`, and `has_more`, returning only the requested node window.
* **Bug Fixes**
  * Improved “no scene open” responses to return a structured empty page with `has_more: false`.
  * Normalized hierarchy response shape to remove the unused `root` field.
  * Safer pagination fallback: `limit=0` means “no limit” and negative `offset` is clamped to `0`.
* **Tests**
  * Expanded unit and integration tests to cover server-side passthrough, fallbacks, and edge cases.

### CODE REVIEW

**Review by dsarno — APPROVED** [review]
Approving — the design is right, the version-skew story is handled in both directions (new-server/old-plugin falls back to server-side slicing; old-server/new-plugin defaults to unlimited), and the test coverage is genuinely good. Two asks before merge, both small:

**1. Medium — `limit <= 0` / negative `offset` mean opposite things on the two paths** (Copilot flagged this too; confirmed). New plugin: `limit=0` → entire tree, `offset` clamped to ≥0. Old-plugin fallback via `paginate()`: `limit=0` → *empty* page with `has_more=True`, and a negative offset slices from the end Python-style. Same tool call, contradictory answers depending on plugin version. Worse, `limit=0` against a new plugin ships the whole tree over the wire — exactly the payload blow-up this PR exists to prevent.

Cleanest fix is to normalize once at the tool boundary in `src/godot_ai/handlers/scene.py` before either path runs: `offset = max(0, offset)`, and either enforce `limit >= 1` in the FastMCP schema (`Field(ge=1)`) or treat `limit <= 0` as unlimited in the fallback to match the plugin's documented semantics.

**2. Low — stale fixture models a response shape no plugin ever sent** (Copilot's second comment; confirmed on the branch). `test_read_scene_hierarchy_resource` (`tests/integration/test_mcp_tools.py:1619-1634`) still mocks the plugin returning `{"root": "Main", ...}` and asserts it round-trips through `godot://scene/hierarchy` — it only passes because the resource handler is a passthrough. Please update the mock to the new contract (`nodes` + `total_count`/`offset`/`limit`/`has_more`) and assert on `nodes[0]`. The fabricated `root` key also lingers in the mocks at `test_mcp_tools.py:67`, `test_mcp_tools.py:89`, and `test_pagination.py:38` — harmless there (the fallback drops it), but worth sweeping in the same pass.

Verified along the way: windowing math is correct including `has_more` at exact boundaries and offset-past-end; the `depth` interaction matches the old behavior (nodes beyond `max_depth` excluded from both list and count); old plugins ignore the new params gracefully; the `godot://scene/hierarchy` resource still reads the full tree via the `limit=0` default; and dropping `root` contradicts nothing documented (tool docstring and `docs/TOOLS.md` never mention it).

🤖 Review drafted with [Claude Code](https://claude.com/claude-code)

File: src/godot_ai/handlers/scene.py [hunk 1]
```diff
@@ -13,9 +13,18 @@ async def scene_get_hierarchy(
...
+    )
+    ## Newer plugins paginate server-side (only the window is walked into node
+    ## dicts and shipped) and stamp `has_more`; pass their result straight
+    ## through. Older plugins — and version-skewed installs — ignore the
+    ## offset/limit params and return the full node list without pagination
+    ## metadata, so fall back to slicing here. This keeps a mixed
+    ## new-server/old-plugin pair correct.
+    if "has_more" in result:
+        return result
+    return paginate(result.get("nodes", []), offset, limit, key="nodes")
```
[the remark this answers is not in the document]
Comment (eth0net) [reply src/godot_ai/handlers/scene.py]: Fixed in a36191a — the old-plugin fallback now clamps a negative `offset` to 0 and treats `limit <= 0` as "no limit" (returns everything from `offset`), matching the plugin contract instead of `paginate()`'s empty-slice / slice-from-end behavior. Added tests for both the `limit=0` and negative-offset cases.

File: plugin/addons/godot_ai/handlers/scene_handler.gd [hunk 2]
```diff
@@ -16,13 +16,38 @@ func _init(connection: McpConnection = null) -> void:
...
+	# accurate total_count, but only materializes those inside the window.
+	var index_ref: Array[int] = [0]
+	_walk_tree(scene_root, nodes, 0, max_depth, scene_root, offset, limit, index_ref)
+	var total: int = index_ref[0]
+	return {"data": {
+		"nodes": nodes,
+		"total_count": total,
+		"offset": offset,
+		"limit": limit,
+		"has_more": limit > 0 and offset + limit < total,
```
[the remark this answers is not in the document]
Comment (eth0net) [reply plugin/addons/godot_ai/handlers/scene_handler.gd:49]: Fixed in a36191a — refreshed `test_read_scene_hierarchy_resource` so its mock sends the current plugin shape (paginated, no `root`) and asserts via `nodes[0].name` + `total_count`/`has_more` rather than the removed `root`. And in 779bbfd the resource now routes through `scene_get_hierarchy` so its shape is consistent across plugin versions.

**Comment by eth0net** [comment]
@coderabbitai good catch on the resource shape — addressed in 779bbfd. `scene_hierarchy_resource_data` now routes through `scene_get_hierarchy` with `limit=0` (full tree) instead of calling the plugin directly, so the `godot://scene/hierarchy` resource returns the same paginated shape (`total_count`/`offset`/`limit`/`has_more`) regardless of plugin version — the old-plugin fallback normalizes it. Strengthened the handler test to assert that shape.

**Review by dsarno — APPROVED** [review]
Thank you — this is the best-engineered PR of the current batch, and it verified clean at every layer.

What makes it good:
- The DFS still visits every in-depth node for an accurate `total_count` but only materializes the `[offset, offset+limit)` window — the per-node dict build and the O(depth) `McpScenePath.from_node` cost are skipped outside it, which is exactly the cost pagination should avoid.
- Version skew is handled in both directions: an old plugin ignoring `offset`/`limit` triggers the Python fallback slice (with matching negative-offset clamp and `limit<=0` = unlimited semantics, both tested), while a new plugin's paginated result passes through untouched via the `has_more` sentinel.
- Dropping the dead `root` field is safe — `handlers/scene.py` was its only consumer and the real plugin never sent it, so it was always `""` in production.

Verification performed on this branch:
- Full Python suite: 1376 passed.
- Full in-editor GDScript suite against a live headless editor: 1873/1893, 0 failed.
- Live end-to-end smoke over the real MCP → WebSocket → editor path with `main.tscn` open: `limit=2` → `[Main, Camera3D]` (`total_count=31`, `has_more=true`); `offset=2,limit=2` → `[DirectionalLight3D, World]` (DFS order continues exactly where the first page ended); `offset=999` → empty window, `has_more=false`, `total_count` intact.

---
_Generated by [Claude Code](https://claude.ai/code)_
